### ISSUE TITLE
Is there any way to get the ip of the client?

### ISSUE THREAD

**Issue by Agoni-sudo** [moquette-io/moquette#867]
Is there any way to get the ip of the client?

**Comment by Yunustt** [comment]
((InetSocketAddress)channel.remoteAddress()).getHostString()

**Comment by Agoni-sudo** [comment]
> （（InetSocketAddress）channel.remoteAddress（））.getHostString（）

Thanks, but how do I get the connected channel

**Comment by Tong-07** [comment]
![Uploading image.png…]()

**Comment by Yunustt** [comment]
There are channels in MQTTConnection

**Comment by Tong-07** [comment]
NewNettyMQTTHandler uses ChannelHandlerContext to create an MQTTConnection
![image](https://github.com/user-attachments/assets/f414e602-149d-4d1b-8654-7ed026d1ca70)

**Comment by Yunustt** [comment]
NewNettyMQTTHandler uses ChannelHandlerContext to create an MQTTConnection
![image](https://github.com/user-attachments/assets/1db20206-4737-4fd4-abac-34180e481324)

**Comment by Agoni-sudo** [comment]
> NewNettyMQTTHandler uses ChannelHandlerContext to create an MQTTConnection ![image](https://private-user-images.githubusercontent.com/90017934/387054990-1db20206-4737-4fd4-abac-34180e481324.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE5MTkzMjIsIm5iZiI6MTczMTkxOTAyMiwicGF0aCI6Ii85MDAxNzkzNC8zODcwNTQ5OTAtMWRiMjAyMDYtNDczNy00ZmQ0LWFiYWMtMzQxODBlNDgxMzI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE4VDA4MzcwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxN2JiMTI2ZDk5YmU1YjFkNjM4YWFjNDBhZDJhMjMyZmI5ZDY4YjAxNGRjZDNjNWU3MzFhZTU0YzBjYzBiYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yz8bEytPn0DsojHhVmu2OrL8db3VlzQh-nS2-Dvu5cU)

Thank you. Is this part of the moquette source code? I want to get the ip in my moquette interceptor. The first time I use moquette, I don't know if it's the wrong way. The way you said to get the ip through the channel ((InetSocketAddress)channel.remoteAddress()).getHostString() should be fine, but this channel seems to be private and I can't call it
@Override
    public void onConnect( InterceptConnectMessage msg) {
        
      
        Connect connect = new Connect();
        connect.setClient_id(msg.getClientID());
        connect.setUsername(msg.getUsername() != null ? msg.getUsername() : "unset");
        connect.setStatus(true);  
        connect.setKeeplive(msg.getKeepAlive());
        connect.setCleansession(msg.isCleanSession());
        connect.setDatetime(new Timestamp(System.currentTimeMillis()));

        log.info("client connect：{}", connect);

        connectService.handleClientConnection(connect);
    }

**Comment by andsel** [comment]
@Agoni-sudo in the interceptor interface you can grab all the information present in `InterceptConnectMessage`, and actually it doesn't contains such information.
Please create an issue describing the desired behavior and label it with `enhancement` 🙏 and close this.

**Comment by Yunustt** [comment]
Use connection callbacks as an example，You can construct an ip property in an InterceptConnectMessage

**Comment by Yunustt** [comment]
@Override
    public void notifyClientConnected(final MqttConnectMessage msg, String remoteAddress) {
        for (final InterceptHandler handler : this.handlers.get(InterceptConnectMessage.class)) {
            LOG.debug("Sending MQTT CONNECT message to interceptor. CId={}, interceptorId={}",
                    msg.payload().clientIdentifier(), handler.getID());
            executor.execute(() -> handler.onConnect(new InterceptConnectMessage(msg, remoteAddress)));
        }
    }

**Comment by andsel** [comment]
Yep, @Yunustt if you or @Agoni-sudo desire to create a PR for this aspect I'll be happy to review :-)

**Comment by robinwu2008** [comment]
My requirement is to determine the source of the client based on the IP address, so if I can get the IP address in IAuthenticator, it would be a perfect solution.

**Comment by HTHou** [comment]
Hi @andsel, I opened #946 to expose the Netty `Channel` through `InterceptConnectMessage`, together with helpers for the client address and port. This addresses the interceptor use case described here while preserving the existing API. Could you please review it when you have a chance? Thanks!
